File: artifacts/imalat-poz-analizleri/constants/kesif.ts
import { PozAnaliz, hesaplaAnalizToplam } from "@/constants/pozAnalizleri";

export interface KesifSatiri {
  id: string;
  analizId: string;
  pozNo: string;
  analizAdi: string;
  olcuBirimi: string;
  birimFiyati: number;
  miktar: number;
  tutar: number;
}

export interface KesifProject {
  id: string;
  ad: string;
  aciklama: string;
  satirlar: KesifSatiri[];
  olusturmaTarihi: string;
  guncellemeTarihi: string;
}

export function genKesifId(prefix: string): string {
  return prefix + Date.now().toString(36) + Math.random().toString(36).slice(2, 6);
}

export function hesaplaSatirTutar(miktar: number, birimFiyati: number): number {
  return Math.round(miktar * birimFiyati * 100) / 100;
}

export function buildKesifSatiri(analiz: PozAnaliz, miktar: number): KesifSatiri {
  const birimFiyati = hesaplaAnalizToplam(analiz).birimFiyati;
  const qty = Number.isFinite(miktar) ? miktar : 0;
  return {
    id: genKesifId("ks"),
    analizId: analiz.id,
    pozNo: analiz.pozNo,
    analizAdi: analiz.analizAdi,
    olcuBirimi: analiz.olcuBirimi,
    birimFiyati,
    miktar: qty,
    tutar: hesaplaSatirTutar(qty, birimFiyati),
  };
}

export function hesaplaKesifToplam(satirlar: KesifSatiri[]): number {
  return Math.round(satirlar.reduce((s, r) => s + r.tutar, 0) * 100) / 100;
}

export function trFmtKesif(n: number): string {
  return n.toLocaleString("tr-TR", { minimumFractionDigits: 2, maximumFractionDigits: 2 });
}

export function sanitizeKesifProjects(stored: unknown): KesifProject[] {
  if (!Array.isArray(stored)) return [];
  const result: KesifProject[] = [];
  for (const raw of stored) {
    if (!raw || typeof raw !== "object") continue;
    const p = raw as Partial<KesifProject>;
    if (typeof p.id !== "string" || typeof p.ad !== "string") continue;
    const satirlar: KesifSatiri[] = [];
    if (Array.isArray(p.satirlar)) {
      for (const s of p.satirlar) {
        if (!s || typeof s !== "object") continue;
        const row = s as Partial<KesifSatiri>;
        if (typeof row.id !== "string" || typeof row.analizId !== "string") continue;
        const miktar = Number(row.miktar) || 0;
        const birimFiyati = Number(row.birimFiyati) || 0;
        satirlar.push({
          id: row.id,
          analizId: row.analizId,
          pozNo: String(row.pozNo ?? ""),
          analizAdi: String(row.analizAdi ?? ""),
          olcuBirimi: String(row.olcuBirimi ?? ""),
          birimFiyati,
          miktar,
          tutar: hesaplaSatirTutar(miktar, birimFiyati),
        });
      }
    }
    result.push({
      id: p.id,
      ad: p.ad.trim() || "Keşif",
      aciklama: typeof p.aciklama === "string" ? p.aciklama : "",
      satirlar,
      olusturmaTarihi:
        typeof p.olusturmaTarihi === "string" ? p.olusturmaTarihi : new Date().toISOString(),
      guncellemeTarihi:
        typeof p.guncellemeTarihi === "string" ? p.guncellemeTarihi : new Date().toISOString(),
    });
  }
  return result.sort(
    (a, b) => new Date(b.guncellemeTarihi).getTime() - new Date(a.guncellemeTarihi).getTime(),
  );
}

File: artifacts/santijet-bfa-flutter/lib/domain/entities/kesif.dart
import 'package:equatable/equatable.dart';

import '../../core/utils/id_gen.dart';
import '../calc/analiz_hesap.dart';
import '../enums/app_enums.dart';
import 'metraj_kalemi.dart';
import 'poz_analiz.dart';

export 'metraj_kalemi.dart';

/// Keşif satırındaki birim fiyatın kaynağı.
enum KesifFiyatKaynagi {
  analiz,
  katalog,
  manuel;

  String get label => switch (this) {
        KesifFiyatKaynagi.analiz => 'Analiz',
        KesifFiyatKaynagi.katalog => 'Katalog',
        KesifFiyatKaynagi.manuel => 'Manuel',
      };

  static KesifFiyatKaynagi fromJson(String? raw) => switch (raw) {
        'analiz' => KesifFiyatKaynagi.analiz,
        'manuel' => KesifFiyatKaynagi.manuel,
        _ => KesifFiyatKaynagi.katalog,
      };

  String get jsonValue => name;
}

/// Keşif satırı — poz + metraj cetveli kalemleri + fiyat.
class KesifSatiri extends Equatable {
  const KesifSatiri({
    required this.id,
    required this.analizId,
    required this.pozNo,
    required this.analizAdi,
    required this.olcuBirimi,
    required this.birimFiyati,
    required this.miktar,
    required this.tutar,
    this.fiyatKaynagi = KesifFiyatKaynagi.katalog,
    this.metrajNotu = '',
    this.discipline = AnalizDiscipline.insaat,
    this.metrajKalemleri = const [],
  });

  final String id;
  final String analizId;
  final String pozNo;
  final String analizAdi;
  final String olcuBirimi;
  final double birimFiyati;
  final double miktar;
  final double tutar;
  final KesifFiyatKaynagi fiyatKaynagi;
  final String metrajNotu;
  final AnalizDiscipline discipline;
  final List<MetrajKalemi> metrajKalemleri;

  /// Cetvel kalemleri varsa toplamları; yoksa elle girilen miktar.
  double get hesaplananMetraj {
    if (metrajKalemleri.isEmpty) return miktar;
    return metrajKalemleri.fold<double>(0, (s, k) => s + k.miktar);
  }

  KesifSatiri withMetrajRollup() {
    final m = hesaplananMetraj;
    return copyWith(
      miktar: m,
      tutar: AnalizHesap.satirTutar(m, birimFiyati),
    );
  }

  KesifSatiri copyWith({
    String? id,
    String? analizId,
    String? pozNo,
    String? analizAdi,
    String? olcuBirimi,
    double? birimFiyati,
    double? miktar,
    double? tutar,
    KesifFiyatKaynagi? fiyatKaynagi,
    String? metrajNotu,
    AnalizDiscipline? discipline,
    List<MetrajKalemi>? metrajKalemleri,
  }) {
    return KesifSatiri(
      id: id ?? this.id,
      analizId: analizId ?? this.analizId,
      pozNo: pozNo ?? this.pozNo,
      analizAdi: analizAdi ?? this.analizAdi,
      olcuBirimi: olcuBirimi ?? this.olcuBirimi,
      birimFiyati: birimFiyati ?? this.birimFiyati,
      miktar: miktar ?? this.miktar,
      tutar: tutar ?? this.tutar,
      fiyatKaynagi: fiyatKaynagi ?? this.fiyatKaynagi,
      metrajNotu: metrajNotu ?? this.metrajNotu,
      discipline: discipline ?? this.discipline,
      metrajKalemleri: metrajKalemleri ?? this.metrajKalemleri,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'analizId': analizId,
        'pozNo': pozNo,
        'analizAdi': analizAdi,
        'olcuBirimi': olcuBirimi,
        'birimFiyati': birimFiyati,
        'miktar': miktar,
        'tutar': tutar,
        'fiyatKaynagi': fiyatKaynagi.jsonValue,
        'metrajNotu': metrajNotu,
        'discipline': discipline.jsonValue,
        'metrajKalemleri': metrajKalemleri.map((k) => k.toJson()).toList(),
      };

  factory KesifSatiri.fromJson(Map<dynamic, dynamic> json) {
    final miktar = (json['miktar'] as num?)?.toDouble() ?? 0;
    final birimFiyati = (json['birimFiyati'] as num?)?.toDouble() ?? 0;
    final pozNo = json['pozNo'] as String? ?? '';
    final rawKalem = json['metrajKalemleri'];
    final kalemler = rawKalem is List
        ? rawKalem
            .whereType<Map<dynamic, dynamic>>()
            .map(MetrajKalemi.fromJson)
            .where((k) => k.id.isNotEmpty)
            .toList()
        : const <MetrajKalemi>[];
    final discipline = json['discipline'] != null
        ? AnalizDiscipline.fromJson(json['discipline'] as String?)
        : AnalizDiscipline.fromPozNo(pozNo);
    var row = KesifSatiri(
      id: json['id'] as String? ?? '',
      analizId: json['analizId'] as String? ?? '',
      pozNo: pozNo,
      analizAdi: json['analizAdi'] as String? ?? '',
      olcuBirimi: json['olcuBirimi'] as String? ?? '',
      birimFiyati: birimFiyati,
      miktar: miktar,
      tutar: AnalizHesap.satirTutar(miktar, birimFiyati),
      fiyatKaynagi: KesifFiyatKaynagi.fromJson(json['fiyatKaynagi'] as String?),
      metrajNotu: json['metrajNotu'] as String? ?? '',
      discipline: discipline,
      metrajKalemleri: kalemler,
    );
    if (kalemler.isNotEmpty) row = row.withMetrajRollup();
    return row;
  }

  @override
  List<Object?> get props => [
        id,
        analizId,
        pozNo,
        analizAdi,
        olcuBirimi,
        birimFiyati,
        miktar,
        tutar,
        fiyatKaynagi,
        metrajNotu,
        discipline,
        metrajKalemleri,
      ];
}

/// Keşif projesi — React Native `KesifProject` arayüzünün karşılığı.
class KesifProject extends Equatable {
  const KesifProject({
    required this.id,
    required this.ad,
    required this.aciklama,
    required this.satirlar,
    required this.olusturmaTarihi,
    required this.guncellemeTarihi,
    this.kod = '',
    this.konum = '',
  });

  final String id;
  final String ad;
  final String aciklama;
  final List<KesifSatiri> satirlar;
  final String olusturmaTarihi;
  final String guncellemeTarihi;
  final String kod;
  final String konum;

  double get toplam => satirlar.fold<double>(0, (sum, row) => sum + row.tutar);

  Map<String, double> get toplamByOlcuBirimi {
    final map = <String, double>{};
    for (final s in satirlar) {
      final key = s.olcuBirimi.trim().isEmpty ? 'Diğer' : s.olcuBirimi.trim();
      map[key] = (map[key] ?? 0) + s.tutar;
    }
    return map;
  }

  /// Disipline göre gruplanmış satırlar (boş disiplinler atlanır).
  Map<AnalizDiscipline, List<KesifSatiri>> get satirlarByDiscipline {
    final map = <AnalizDiscipline, List<KesifSatiri>>{
      for (final d in AnalizDiscipline.kesifSirasi) d: <KesifSatiri>[],
    };
    for (final s in satirlar) {
      map.putIfAbsent(s.discipline, () => <KesifSatiri>[]).add(s);
    }
    return map;
  }

  KesifProject copyWith({
    String? id,
    String? ad,
    String? aciklama,
    List<KesifSatiri>? satirlar,
    String? olusturmaTarihi,
    String? guncellemeTarihi,
    String? kod,
    String? konum,
  }) {
    return KesifProject(
      id: id ?? this.id,
      ad: ad ?? this.ad,
      aciklama: aciklama ?? this.aciklama,
      satirlar: satirlar ?? this.satirlar,
      olusturmaTarihi: olusturmaTarihi ?? this.olusturmaTarihi,
      guncellemeTarihi: guncellemeTarihi ?? this.guncellemeTarihi,
      kod: kod ?? this.kod,
      konum: konum ?? this.konum,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'ad': ad,
        'aciklama': aciklama,
        'satirlar': satirlar.map((s) => s.toJson()).toList(),
        'olusturmaTarihi': olusturmaTarihi,
        'guncellemeTarihi': guncellemeTarihi,
        'kod': kod,
        'konum': konum,
      };

  factory KesifProject.fromJson(Map<dynamic, dynamic> json) {
    final rawSatirlar = json['satirlar'];
    return KesifProject(
      id: json['id'] as String? ?? '',
      ad: (json['ad'] as String? ?? 'Keşif').trim(),
      aciklama: json['aciklama'] as String? ?? '',
      satirlar: rawSatirlar is List
          ? rawSatirlar
              .whereType<Map<dynamic, dynamic>>()
              .map(KesifSatiri.fromJson)
              .where((s) => s.id.isNotEmpty)
              .toList()
          : const [],
      olusturmaTarihi: json['olusturmaTarihi'] as String? ??
          DateTime.now().toIso8601String(),
      guncellemeTarihi: json['guncellemeTarihi'] as String? ??
          DateTime.now().toIso8601String(),
      kod: (json['kod'] as String? ?? '').trim().toUpperCase(),
      konum: (json['konum'] as String? ?? '').trim(),
    );
  }

  @override
  List<Object?> get props => [
        id,
        ad,
        aciklama,
        satirlar,
        olusturmaTarihi,
        guncellemeTarihi,
        kod,
        konum,
      ];
}

KesifSatiri buildKesifSatiri(PozAnaliz analiz, double miktar) {
  final hesap = AnalizHesap.hesapla(analiz);
  final birimFiyati =
      hesap.birimFiyati > 0 ? hesap.birimFiyati : analiz.birimFiyati;
  final kaynak = analiz.kaynakTip == KaynakTip.sistem
      ? KesifFiyatKaynagi.katalog
      : KesifFiyatKaynagi.analiz;
  final discipline =
      analiz.discipline ?? AnalizDiscipline.fromPozNo(analiz.pozNo);
  return KesifSatiri(
    id: IdGen.make('ks'),
    analizId: analiz.id,
    pozNo: analiz.pozNo,
    analizAdi: analiz.analizAdi,
    olcuBirimi: analiz.olcuBirimi,
    birimFiyati: birimFiyati,
    miktar: miktar.isFinite ? miktar : 0,
    tutar: AnalizHesap.satirTutar(miktar, birimFiyati),
    fiyatKaynagi: kaynak,
    discipline: discipline,
  );
}
